### PR TITLE
chore: Auth module revamp

### LINKED ISSUE #4315: Auth-module Revamp
Technical improvements to the auth module, moving API endpoints to server actions, code rewrite and optimizations with move to new `modules` folder and splitting auth features into community and enterprise features.

Full Ticket: https://www.notion.so/formbricks/Auth-module-Revamp-13ee5de84a588049abdac744aae90c73?pvs=4

### PR DESCRIPTION
**Pull request by Dhruwang**
## What does this PR do?

Fixes #4315 

Relocates and refactors auth code. 
Makes 2fa and SSO enterprise features

## How should this be tested?

Test auth, two factor , sso

### Required

- [x] Filled out the "How to test" section in this PR
- [x] Read [How we Code at Formbricks](<[https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md](https://formbricks.com/docs/contributing/how-we-code)>)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand bits
- [x] Ran `pnpm build`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues
- [x] First PR at Formbricks? [Please sign the CLA!](https://cla-assistant.io/formbricks/formbricks) Without it we wont be able to merge it 🙏

### Appreciated

- [x] If a UI change was made: Added a screen recording or screenshots to this PR
- [x] Updated the Formbricks Docs if changes were necessary

## Release Notes

- **New Features**
  - Enhanced documentation for configuring Formbricks with external authentication providers, including SSO functionality.
  - Introduced a new `ForgotPasswordForm` component for password reset requests.
  - Added a `ResetPasswordForm` component for user password resets.
  - New `SSOOptions` component for displaying various Single Sign-On options.
  - Added `EmailSentPage` and `ForgotPasswordPage` components for improved user experience in password recovery.
  - Introduced `SignupWithoutVerificationSuccessPage` for successful sign-up notifications.
  - Added `DisplayBackupCodes` and `EnableTwoFactorModal` components to enhance two-factor authentication setup.
  - Updated `ConfirmPasswordForm`, `ScanQRCode`, and `EnterCode` components to facilitate two-factor authentication.
  
- **Improvements**
  - Improved error handling and validation for user-related actions.
  - Enhanced localization support with new translation keys for various languages.
  - Streamlined user management functions and improved feedback mechanisms during authentication processes.
  - Updated documentation for clarity on Google OAuth integration and other authentication methods.

- **Bug Fixes**
  - Resolved issues with user feedback during authentication and verification processes.

- **Chores**
  - Removed deprecated components and streamlined authentication flows.

### CODE REVIEW

File: apps/web/modules/ee/two-factor-auth/components/disable-two-factor-modal.tsx [hunk 1]
```diff
@@ -3,65 +3,71 @@
...
     const { code, password, backupCode } = data;
-
     const disableTwoFactorAuthResponse = await disableTwoFactorAuthAction({ code, password, backupCode });
-
     if (disableTwoFactorAuthResponse?.data) {
       toast.success(disableTwoFactorAuthResponse.data.message);
-
-      router.refresh();
-      resetState();
+      form.reset();
```
Comment (pandeymangg) [inline apps/web/modules/ee/two-factor-auth/components/disable-two-factor-modal.tsx:54]: I think we should leave the `router.refresh` function call here to sync the server state again

**Review by pandeymangg — CHANGES_REQUESTED** [review]
great job on the PR @Dhruwang  already looks pretty good, I just added a few comments on the PR, please go through them, and also implement the new lib approach that we talked about on slack, thanks a lot 🙌

File: packages/types/user.ts [hunk 2]
```diff
@@ -28,12 +28,21 @@ export const ZUserName = z
   .min(1, { message: "Name should be at least 1 character long" })
   .regex(/^[\p{L}\p{M}\s'\d-]+$/u, "Invalid name format");
 
+export const ZUserEmail = z.string().email();
+
+export const ZUserPassword = z
```
[the remark this answers is not in the document]
Comment (Dhruwang) [reply packages/types/user.ts:33]: Screenshot 2024-11-22 at 2 02 54 PM

File: apps/web/modules/auth/forgot-password/actions.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import { actionClient } from "@/lib/utils/action-client";
+import { sendForgotPasswordEmail } from "@/modules/email";
+import { z } from "zod";
+import { getUserByEmail } from "@formbricks/lib/user/service";
+
+const ZForgotPasswordAction = z.object({
+  email: z.string().max(255).email({ message: "Invalid email" }),
+});
+
+export const forgotPasswordAction = actionClient
```
[the remark this answers is not in the document]
Comment (Dhruwang) [reply apps/web/modules/auth/forgot-password/actions.ts:12]: Here we removed error throwing few days back as a part of security fix, according to which user should not be able to find out which emails are registered and which arent on platform

**Review by pandeymangg — APPROVED** [review]
LGTM! 🚀

**Review by mattinannt — APPROVED** [review]

**Pull request merged**